pkg/pool-stable-phantom/test/StablePoolRates.test.ts

```diff
@@ -0,0 +1,811 @@
+
+            const rateProvider = await deployedAt('v2-pool-utils/MockRateProvider', allRateProviders[i]);
+            await rateProvider.mockRate(fp(1 - i / 10));
+            await pool.updateTokenRateCache(token.address);
+          });
+          expectedScalingFactors = await getExpectedScalingFactors(allTokens, await getRates(allRateProviders));
+
+          // Set rate to zero. If the pool is reading from the rate provider directly then this will cause reverts.
+          // This ensures that the pool is using it's cache properly.
+          await allTokens.asyncEach(async (_, i) => {
+            if (allRateProviders[i] === ZERO_ADDRESS) return;
+
+            const rateProvider = await deployedAt('v2-pool-utils/MockRateProvider', allRateProviders[i]);
+            await rateProvider.mockRate(fp(0));
+          });
+        });
+
+        itAdaptsTheScalingFactorsCorrectly();
+      });
+    });
+
+    describe('getAdjustedBalances', () => {
+      let allTokens: TokenList;
+      let allRateProviders: string[];
+      let allExemptFlags: boolean[];
+      let rates: BigNumber[];
+
+      sharedBeforeEach('deploy pool', async () => {
+        await deployPoolSimple(owner, tokens);
+        allTokens = await tokensWithBpt();
+        allRateProviders = await rateProvidersWithBpt();
+        allExemptFlags = await exemptFlagsWithBpt();
+      });
+
+      sharedBeforeEach('mock rates', async () => {
+        await allTokens.asyncEach(async (token, i) => {
+          if (allRateProviders[i] === ZERO_ADDRESS) return;
+          const rateProvider = await deployedAt('v2-pool-utils/MockRateProvider', allRateProviders[i]);
+          await rateProvider.mockRate(fp(1 + i / 10));
+          await pool.updateTokenRateCache(token.address);
+        });
+        rates = await getRates(allRateProviders);
+
+        // Set rates to zero. If the pool is reading from the rate provider directly then this will cause reverts.
+        // This ensures that the pool is using its cache properly.
+        await allTokens.asyncEach(async (_, i) => {
+          if (allRateProviders[i] === ZERO_ADDRESS) return;
+
+          const rateProvider = await deployedAt('v2-pool-utils/MockRateProvider', allRateProviders[i]);
+          await rateProvider.mockRate(fp(0));
+        });
+      });
+
+      context('when ignoring exempt flags', () => {
+        it('returns the array  with elements scaled by the cached token rate', async () => {
+          for (let i = 0; i < 5; i++) {
+            const inputArray = allTokens.map(() => fp(Math.random()));
+            const expectedOutputArray = inputArray.map((input, i) => input.mul(fp(1)).div(rates[i]));
+
+            expect(await pool.getAdjustedBalances(inputArray, true)).to.be.deep.eq(expectedOutputArray);
+          }
+        });
+      });
+
+      context('when not ignoring exempt flags', () => {
+        it('returns the array with elements scaled by the cached token rate if exempt', async () => {
```

Comment:
```suggestion
        it('returns the array  with elements scaled by the old token rate', async () => {
          for (let i = 0; i < 5; i++) {
            const inputArray = allTokens.map(() => fp(Math.random()));
            const expectedOutputArray = inputArray.map((input, i) => input.mul(fp(1)).div(rates[i]));

            expect(await pool.getAdjustedBalances(inputArray, true)).to.be.deep.eq(expectedOutputArray);
          }
        });
      });

      context('when not ignoring exempt flags', () => {
        it('returns the array with elements scaled by the old token rate if exempt', async () => {
```

Reply:
I've update this to make clear that it's referring to a ratio.